### PR TITLE
fix(pacmak): .NET bindings fail to compile with error CS8120

### LINKED ISSUE #3759: C# Runtime check fails on compilation
### Describe the bug

The CDKTF library does not build with 1.67.0 after upgrading from 1.65.0 in C#. There is an error thrown when running `dotnet build --force --configuration Release` as part of `jsii-pacmak` as part of `jsii-srcmak`.

### Expected Behavior

The unchanged code should still build.

### Current Behavior

```
[jsii-pacmak] [WARN] Exception occurred, not cleaning up /var/folders/z_/v03l33d55fb57nrr3b1q03ch0000gq/T/npm-packP6yknl
[jsii-pacmak] [WARN] dotnet failed
Error: All attempts failed. Last error: Command (dotnet build --force --configuration Release) failed with status 1:

/private/var/folders/z_/v03l33d55fb57nrr3b1q03ch0000gq/T/npm-pack9gX9b6/HashiCorp.Cdktf/HashiCorp/Cdktf/ListTerraformIterator.cs(33,26): error CS8120: The switch case is unreachable. It has already been handled by a previous case or it is impossible to match. [/private/var/folders/z_/v03l33d55fb57nrr3b1q03ch0000gq/T/npm-pack9gX9b6/HashiCorp.Cdktf/HashiCorp.Cdktf.csproj]
/private/var/folders/z_/v03l33d55fb57nrr3b1q03ch0000gq/T/npm-pack9gX9b6/HashiCorp.Cdktf/HashiCorp/Cdktf/ListTerraformIterator.cs(35,26): error CS8120: The switch case is unreachable. It has already been handled by a previous case or it is impossible to match. [/private/var/folders/z_/v03l33d55fb57nrr3b1q03ch0000gq/T/npm-pack9gX9b6/HashiCorp.Cdktf/HashiCorp.Cdktf.csproj]
/private/var/folders/z_/v03l33d55fb57nrr3b1q03ch0000gq/T/npm-pack9gX9b6/HashiCorp.Cdktf/HashiCorp/Cdktf/ListTerraformIterator.cs(37,26): error CS8120: The switch case is unreachable. It has already been handled by a previous case or it is impossible to match. [/private/var/folders/z_/v03l33d55fb57nrr3b1q03ch0000gq/T/npm-pack9gX9b6/HashiCorp.Cdktf/HashiCorp.Cdktf.csproj]
/private/var/folders/z_/v03l33d55fb57nrr3b1q03ch0000gq/T/npm-pack9gX9b6/HashiCorp.Cdktf/HashiCorp/Cdktf/ListTerraformIterator.cs(39,26): error CS8120: The switch case is unreachable. It has already been handled by a previous case or it is impossible to match. [/private/var/folders/z_/v03l33d55fb57nrr3b1q03ch0000gq/T/npm-pack9gX9b6/HashiCorp.Cdktf/HashiCorp.Cdktf.csproj]
/private/var/folders/z_/v03l33d55fb57nrr3b1q03ch0000gq/T/npm-pack9gX9b6/HashiCorp.Cdktf/HashiCorp/Cdktf/ListTerraformIterator.cs(41,26): error CS8120: The switch case is unreachable. It has already been handled by a previous case or it is impossible to match. [/private/var/folders/z_/v03l33d55fb57nrr3b1q03ch0000gq/T/npm-pack9gX9b6/HashiCorp.Cdktf/HashiCorp.Cdktf.csproj]
/private/var/folders/z_/v03l33d55fb57nrr3b1q03ch0000gq/T/npm-pack9gX9b6/HashiCorp.Cdktf/HashiCorp/Cdktf/TerraformIterator.cs(57,26): error CS8120: The switch case is unreachable. It has already been handled by a previous case or it is impossible to match. [/private/var/folders/z_/v03l33d55fb57nrr3b1q03ch0000gq/T/npm-pack9gX9b6/HashiCorp.Cdktf/HashiCorp.Cdktf.csproj]
/private/var/folders/z_/v03l33d55fb57nrr3b1q03ch0000gq/T/npm-pack9gX9b6/HashiCorp.Cdktf/HashiCorp/Cdktf/TerraformIterator.cs(59,26): error CS8120: The switch case is unreachable. It has already been handled by a previous case or it is impossible to match. [/private/var/folders/z_/v03l33d55fb57nrr3b1q03ch0000gq/T/npm-pack9gX9b6/HashiCorp.Cdktf/HashiCorp.Cdktf.csproj]
/private/var/folders/z_/v03l33d55fb57nrr3b1q03ch0000gq/T/npm-pack9gX9b6/HashiCorp.Cdktf/HashiCorp/Cdktf/TerraformIterator.cs(61,26): error CS8120: The switch case is unreachable. It has already been handled by a previous case or it is impossible to match. [/private/var/folders/z_/v03l33d55fb57nrr3b1q03ch0000gq/T/npm-pack9gX9b6/HashiCorp.Cdktf/HashiCorp.Cdktf.csproj]
/private/var/folders/z_/v03l33d55fb57nrr3b1q03ch0000gq/T/npm-pack9gX9b6/HashiCorp.Cdktf/HashiCorp/Cdktf/TerraformIterator.cs(63,26): error CS8120: The switch case is unreachable. It has already been handled by a previous case or it is impossible to match. [/private/var/folders/z_/v03l33d55fb57nrr3b1q03ch0000gq/T/npm-pack9gX9b6/HashiCorp.Cdktf/HashiCorp.Cdktf.csproj]
/private/var/folders/z_/v03l33d55fb57nrr3b1q03ch0000gq/T/npm-pack9gX9b6/HashiCorp.Cdktf/HashiCorp/Cdktf/TerraformIterator.cs(65,26): error CS8120: The switch case is unreachable. It has already been handled by a previous case or it is impossible to match. [/private/var/folders/z_/v03l33d55fb57nrr3b1q03ch0000gq/T/npm-pack9gX9b6/HashiCorp.Cdktf/HashiCorp.Cdktf.csproj]
```

The code in question looks like this in C#:

```csharp
public class ListTerraformIterator : HashiCorp.Cdktf.TerraformIterator
    {
        /// <remarks>
        /// <strong>Stability</strong>: Experimental
        /// </remarks>
        public ListTerraformIterator(object list): base(_MakeDeputyProps(list))
        {
        }

        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
        private static DeputyProps _MakeDeputyProps(object list)
        {
            if (Amazon.JSII.Runtime.Configuration.RuntimeTypeChecking)
            {
                switch (list)
                {
                    case string[] cast_a33039:
                        break;
                    case HashiCorp.Cdktf.IResolvable cast_a33039:
                        break;
                    case double[] cast_a33039:
                        break;
                    case HashiCorp.Cdktf.ComplexList cast_a33039:
                        break;
                    case HashiCorp.Cdktf.StringMapList cast_a33039:
                        break;
                    case HashiCorp.Cdktf.NumberMapList cast_a33039:
                        break;
                    case HashiCorp.Cdktf.BooleanMapList cast_a33039:
                        break;
                    case HashiCorp.Cdktf.AnyMapList cast_a33039:
                        break;
                    case object[] cast_a33039:
                        for (var __idx_14fecf = 0 ; __idx_14fecf < cast_a33039.Length ; __idx_14fecf++)
                        {
                            switch (cast_a33039[__idx_14fecf])
                            {
                                case bool cast_30b700:
                                    break;
                                case HashiCorp.Cdktf.IResolvable cast_30b700:
                                    break;
                                case Amazon.JSII.Runtime.Deputy.AnonymousObject cast_30b700:
                                    // Not enough information to type-check...
                                    break;
                                case null:
                                    throw new System.ArgumentException($"Expected argument {nameof(list)}[{__idx_14fecf}] to be one of: bool, {typeof(HashiCorp.Cdktf.IResolvable).FullName}; received null", nameof(list));
                                default:
                                    throw new System.ArgumentException($"Expected argument {nameof(list)}[{__idx_14fecf}] to be one of: bool, {typeof(HashiCorp.Cdktf.IResolvable).FullName}; received {cast_a33039[__idx_14fecf].GetType().FullName}", nameof(list));
                            }
                        }
                        break;
                    case Amazon.JSII.Runtime.Deputy.AnonymousObject cast_a33039:
                        // Not enough information to type-check...
                        break;
                    case null:
                        throw new System.ArgumentException($"Expected argument {nameof(list)} to be one of: string[], {typeof(HashiCorp.Cdktf.IResolvable).FullName}, double[], {typeof(HashiCorp.Cdktf.ComplexList).FullName}, {typeof(HashiCorp.Cdktf.StringMapList).FullName}, {typeof(HashiCorp.Cdktf.NumberMapList).FullName}, {typeof(HashiCorp.Cdktf.BooleanMapList).FullName}, {typeof(HashiCorp.Cdktf.AnyMapList).FullName}, object[]; received null", nameof(list));
                    default:
                        throw new System.ArgumentException($"Expected argument {nameof(list)} to be one of: string[], {typeof(HashiCorp.Cdktf.IResolvable).FullName}, double[], {typeof(HashiCorp.Cdktf.ComplexList).FullName}, {typeof(HashiCorp.Cdktf.StringMapList).FullName}, {typeof(HashiCorp.Cdktf.NumberMapList).FullName}, {typeof(HashiCorp.Cdktf.BooleanMapList).FullName}, {typeof(HashiCorp.Cdktf.AnyMapList).FullName}, object[]; received {list.GetType().FullName}", nameof(list));
                }
            }
            return new DeputyProps(new object?[]{list});
        }
```

The errors are thrown from `case HashiCorp.Cdktf.ComplexList cast_a33039:` to `case HashiCorp.Cdktf.AnyMapList cast_a33039:`.

### Reproduction Steps

```sh
git clone git@github.com:hashicorp/terraform-cdk.git
cd terraform-cdk
git checkout automation/yarn-upgrade
yarn
yarn build
yarn package 
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### SDK version used

1.57.0

### Environment details (OS name and version, etc.)

OSX 12.6

### PR DESCRIPTION
**Pull request by RomainMuller**
If a type union includes several candidates that are related to each other (A extends B or A implements B), `jsii-pacmak` may generate type checking clauses in a pattern matching `switch` statement in an order such that the compiler identifies dead clauses, which is an error in C#.

This adds a provision to NOT emit such a clause so as to not cause the error. It is worth mentioning that the error cannot be "opted out" of via a `#pragma` directive like a warning would be, which is unfortunate.

Fixes #3759

---

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license].

[Apache 2.0 license]: https://www.apache.org/licenses/LICENSE-2.0

### CODE REVIEW

File: tsconfig-base.json [hunk 1]
```diff
@@ -21,5 +21,6 @@
     "composite": true,                     /* Ensure TypeScript can determine where to find the outputs of the referenced project to compile project. */
     "incremental": true,                   /* Enable incremental compilation by reading/writing information from prior compilations to a file on disk. */
     "useDefineForClassFields": false,      /* Use Object.defineProperty for introducing class fields. */
+    "skipLibCheck": true,                  /* Skip type checking of declaration files. */
```
Comment (DanielMSchmidt) [inline tsconfig-base.json:24]: This seems unrelated, was this an error?

File: packages/jsii-pacmak/lib/targets/dotnet/runtime-type-checking.ts [hunk 2]
```diff
@@ -172,6 +172,26 @@ abstract class Validation {
         code.openBlock(`switch (${expression})`);
         for (const type of types) {
           validTypes.push(resolver.toDotNetTypeName(type.spec!));
+
+          /**
```
Comment (DanielMSchmidt) [inline packages/jsii-pacmak/lib/targets/dotnet/runtime-type-checking.ts:176]: Great explanation 👍

**Review by DanielMSchmidt — APPROVED** [review]

**Review by iliapolo — APPROVED** [review]

File: tsconfig-base.json (hunk 1 above)
Comment (RomainMuller) [reply tsconfig-base.json:24]: Yeah this is unrelated but does not hurt :) It's a recommended setting in TypeScript and could make builds somewhat faster.

**Comment by RomainMuller** [comment]
@MergifyIO update

**Pull request merged**